packages/transactions/src/serializers/header.ts

```diff
@@ -1,51 +1,104 @@
+import { getStructCodec, getStructDecoder, getStructEncoder } from '@solana/codecs-data-structures';
+import { getU8Codec, getU8Decoder, getU8Encoder } from '@solana/codecs-numbers';
 
 import { getCompiledMessageHeader } from '../compile-header';
 
 type MessageHeader = ReturnType<typeof getCompiledMessageHeader>;
 
-export function getMessageHeaderCodec(): Serializer<MessageHeader> {
-    return struct(
+let memoizedU8Encoder: Encoder<number> | undefined;
+function getMemoizedU8Encoder(): Encoder<number> {
+    if (!memoizedU8Encoder) memoizedU8Encoder = getU8Encoder();
+    return memoizedU8Encoder;
+}
+
+function getMemoizedU8EncoderDescription(description?: string): Encoder<number> {
+    const encoder = getMemoizedU8Encoder();
+    return {
+        ...encoder,
+        description: description ?? encoder.description,
+    };
+}
+
+let memoizedU8Decoder: Decoder<number> | undefined;
+function getMemoizedU8Decoder(): Decoder<number> {
+    if (!memoizedU8Decoder) memoizedU8Decoder = getU8Decoder();
+    return memoizedU8Decoder;
+}
+
+function getMemoizedU8DecoderDescription(description?: string): Decoder<number> {
+    const decoder = getMemoizedU8Decoder();
+    return {
+        ...decoder,
+        description: description ?? decoder.description,
+    };
+}
+
+let memoizedU8Codec: Codec<number> | undefined;
+function getMemoizedU8Codec(): Codec<number> {
+    if (!memoizedU8Codec) memoizedU8Codec = getU8Codec();
+    return memoizedU8Codec;
+}
+
+function getMemoizedU8CodecDescription(description?: string): Codec<number> {
+    const codec = getMemoizedU8Codec();
+    return {
+        ...codec,
+        description: description ?? codec.description,
+    };
+}
+
+const numSignerAccountsDescription = __DEV__
+    ? 'The expected number of addresses in the static address list belonging to accounts that are required to sign this transaction'
+    : undefined;
+
+const numReadonlySignerAccountsDescription = __DEV__
+    ? 'The expected number of addresses in the static address list belonging to accounts that are required to sign this transaction, but may not be writable'
+    : undefined;
+
+const numReadonlyNonSignerAccountsDescription = __DEV__
+    ? 'The expected number of addresses in the static address list belonging to accounts that are neither signers, nor writable'
+    : undefined;
+
+const messageHeaderDescription = __DEV__
+    ? 'The transaction message header containing counts of the signer, readonly-signer, and readonly-nonsigner account addresses'
+    : undefined;
+
+export function getMessageHeaderEncoder(): Encoder<MessageHeader> {
+    return getStructEncoder(
+        [
+            ['numSignerAccounts', getMemoizedU8EncoderDescription(numSignerAccountsDescription)],
+            ['numReadonlySignerAccounts', getMemoizedU8EncoderDescription(numReadonlySignerAccountsDescription)],
+            ['numReadonlyNonSignerAccounts', getMemoizedU8EncoderDescription(numReadonlyNonSignerAccountsDescription)],
```

Comment:
Would it not be better to memoize the entire `MessageHeader(En|De)coder` instead of all their individual components?

Reply:
So far I've taken the approach that the caller should `memoize` as it wants, so the caller of `getMessageHeaderEncoder` etc will memoize it in a following PR. So I'm kind of working my way up the stack adding memoization at the call sites.

I'm not completely sure if this is the best approach. But I thought it seemed wrong to add memoization to the `codecs-strings` functions for example, because I don't think that module should be concerned with memoization of the base codecs. Which leads to its caller in #1726 memoizing eg `getBase58Encoder` from that package. So I thought continuing that pattern was the best way to separate concerns, where the higher-level codecs memoize the lower level ones that they build on, but not themselves.

Reply:
Cool makes sense, I'll just add it where I think it makes sense/where we have more complexity. Agreed that we don't need it everywhere! 

> Ultimately function programming is bound to call lots of functions.

Agree but technically a bunch of functions could be consts if it wasn't for tree shaking 😄